core/core/src/sf_core.rs

```diff
@@ -139,28 +205,62 @@ impl OneClientCore {
     }
 
     pub fn perform(&mut self) -> Result<Result<HostValue, HostValue>, PerformExceptionError> {
+        // we can't send metrics if we don't even know the profile and provider urls
         let perform_input = PerformInput::take_in(MessageExchangeFfi)?;
-        let profile_url = perform_input.profile_url.clone();
-        let provider_url = perform_input.provider_url.clone();
-        
-        let result = self.perform_inner(perform_input);
-        crate::observability::metrics::log_metric!(
-            Perform
-            // success is when the result is `Ok(Ok(_))`
-            success = result.as_ref().map(|inner| inner.is_ok()).unwrap_or(false),
-            profile = &profile_url,
-            provider = &provider_url
-        );
-
-        result
-    }
 
-    /// Internal implementation of perform so we can have ergonomics of `?` operator while handling metrics outside
-    fn perform_inner(&mut self, perform_input: PerformInput) -> Result<Result<HostValue, HostValue>, PerformExceptionError> {
-        self.document_cache.cache(&perform_input.profile_url)?;
-        self.document_cache.cache(&perform_input.provider_url)?;
-        self.document_cache.cache(&perform_input.map_url)?;
+        // information we have so far parsed from the available data, might be partial if an exception happens
+        let mut metrics_data = PerformMetricsData {
+            profile_url: &perform_input.profile_url,
+            provider_url: &perform_input.provider_url,
+            map_url: &perform_input.map_url,
+            ..Default::default()
+        };
+
+        /// This is like the `?` operator but allows us to run cleanup code within the same borrow scope,
+        /// so we don't have to clone metrics data around for no reason
+        macro_rules! try_metrics {
+            (Send $success: expr) => {
+                tracing::debug!(perform_metrics = ?metrics_data);
```

Comment:
Yes, it is sort of like a local lambda in this sense.

Reply:
Since the binding `let mut metrics_data = PerformMetricsData { ... }` is above it it can see this identifier and refer to it.

Reply:
But unlike a lamba, there is no "capturing" going on, so at the macro invocation site the same identifier has to still be accessible. You can't, for example, do this:
```rs
fn outer() {
  let value = false;
  macro_rules! my_macro {
    () => { if value { panic!() } };
  }
  
  fn inner() {
    // this function isn't a closure/lambda, it doesn't capture `value` from the outer scope
    // and so this macro cannot be used here
    
    // even if you did something like this
    let value = true;
    // the macro refers to that one specific `value` binding and since that is not accesible here it will error
    
    my_macro!();
  }
  inner();
  
   // this will work, but will refer to the `value = false` binding that came before the macro definiton, no this one later on
  let value = true;
  my_macro!();
}
```

Reply:
See playground I made to check what I'm saying here: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=5cdd16b227c5e4b2efe7c88a16010228 - if you uncomment line 15 you will get the compilation error (it even tells you you can use a closure to fix it, which is nice)